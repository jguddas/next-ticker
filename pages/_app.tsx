import { ThemeProvider, Preflight } from '@xstyled/styled-components'
import type { AppProps } from 'next/app'
import defaultTheme from '../theme/default'

export default ({ Component, pageProps }: AppProps): JSX.Element => (
  <ThemeProvider theme={defaultTheme}>
    <Preflight />
    <Component {...pageProps} />
  </ThemeProvider>
)
