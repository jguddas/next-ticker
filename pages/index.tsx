import React from 'react'
import dayjs from 'dayjs'
import pMemoize from 'p-memoize'
import * as blaze from 'ts-blaze'
import { x } from '@xstyled/styled-components'
import { stringify } from 'query-string'
import useSWR from 'swr'
import PageWrapper from '../components/PageWrapper'
import Box from '../components/Box'
import Donut from '../components/Donut'
import InfoElement from '../components/InfoElement'
import LineGraph from '../components/LineGraph'
import Form, { State as FormState } from '../components/Form'

const initialState: FormState = {
  symbol: 'AAPL',
  initialInvestment: 80000,
  fromDate: '2020-01-01',
  toDate: '2021-01-01',
}

const calcMaxDrawdown = (data: { adjustedClose: number }[]) => {
  let prevMax = -Infinity
  let maxDd = -Infinity

  for (let i = 0; i < data.length; i += 1) {
    if (data[i].adjustedClose > prevMax) {
      prevMax = data[i].adjustedClose
    }
    const dd = (prevMax - data[i].adjustedClose) / prevMax
    if (dd > maxDd) {
      maxDd = dd
    }
  }

  return maxDd
}

const isDate = (val: string) => dayjs(val).format('YYYY-MM-DD') === val
const isApiResponse = blaze.array(
  blaze.object({
    date: blaze.string().satisfies(isDate),
    adjustedClose: blaze.number(),
  })
)

const fetcher = pMemoize(async (input: RequestInfo, init?: RequestInit) => {
  const data = await fetch(input, init).then((r) => r.json())
  if (!isApiResponse(data)) {
    throw Error('api response does not match type')
  }
  return data.sort((a, b) => (dayjs(b.date).isAfter(dayjs(a.date)) ? -1 : 1))
})

export const Home = (): JSX.Element => {
  const [state, setState] = React.useState(initialState)
  const { data, error } = useSWR(
    `/api/series?${stringify({
      symbol: state.symbol,
      fromDate: state.fromDate,
      toDate: state.toDate,
    })}`,
    fetcher
  )
  const maxDrawDown = React.useMemo(
    () => (data ? calcMaxDrawdown(data) : null),
    [data]
  )
  const startValue = data ? data[0].adjustedClose : null
  const endValue = data ? data[data.length - 1].adjustedClose : null
  const percentGain = (endValue - startValue) / startValue
  return (
    <PageWrapper title="Next Ticker" subHeader="some text about this app">
      <x.div display="flex" flexDirection="column" gap="56px">
        <Form initialState={initialState} onValidChange={setState} />
        {error ? <Box title="Error">{error.toString?.()}</Box> : null}
        {data ? (
          <>
            <Box title="Information">
              <x.div
                display="flex"
                flexDirection={{ _: 'column', sm: 'row' }}
                gap={{ _: '36px', sm: '40px' }}
              >
                <Donut
                  max={100}
                  value={Math.max(0, Math.min(1, percentGain)) * 100}
                  alignSelf={{ _: 'center', sm: 'right' }}
                >
                  <InfoElement alignItems="center" title="Growth">
                    {percentGain > 0 ? '+' : ''}
                    {Math.round(percentGain * 10000) / 100}%
                  </InfoElement>
                </Donut>
                <x.div display="flex" flexDirection="column" gap="32px">
                  <InfoElement title="Total investments">
                    {Math.round(
                      (state.initialInvestment +
                        state.initialInvestment * percentGain) *
                        100
                    ) / 100}
                    $
                  </InfoElement>
                  <InfoElement title="Simple return">
                    {Math.round(state.initialInvestment * percentGain * 100) /
                      100}
                    $
                  </InfoElement>
                </x.div>
                <x.div display="flex" flexDirection="column" gap="32px">
                  <InfoElement title="Maximum drawdown">
                    -{Math.round(maxDrawDown * 10000) / 100}%
                  </InfoElement>
                </x.div>
              </x.div>
            </Box>
            <Box title="History">
              <LineGraph data={data.map((val) => val.adjustedClose)} />
            </Box>
          </>
        ) : null}
      </x.div>
    </PageWrapper>
  )
}

export default Home
