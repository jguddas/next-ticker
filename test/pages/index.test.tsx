import React from 'react'
import { render } from '../testUtils'
import { Home } from '../../pages/index'
import 'jest-styled-components'

describe('Home page', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      <Home
        initialState={{
          symbol: 'AAPL',
          initialInvestment: 80000,
          fromDate: '2020-01-01',
          toDate: '2021-01-01',
        }}
      />,
      {}
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
