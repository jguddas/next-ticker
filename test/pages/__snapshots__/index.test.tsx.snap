// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  .c0.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding-bottom: 56px;
}

.c1.c1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1.c1 undefined {
  width: 750px;
  margin-top: 128px;
  margin-right: 120px;
  margin-left: 120px;
}

.c2.c2 {
  margin-bottom: 56px;
}

.c4.c4 {
  color: text;
  font-size: 1px;
}

.c6.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 5px;
  margin-bottom: 5px;
}

.c6.c6 undefined {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c7.c7 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c8.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c9.c9 {
  color: text;
  font-size: 1px;
  margin-bottom: 3px;
}

.c11.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 5px;
}

.c11.c11 undefined {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c5.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 56px;
}

.c3.c3 {
  color: heading;
  font-size: heading;
  line-height: heading;
}

.c10.c10 {
  -webkit-transition: border-color .15s ease-in-out,box-shadow .15s ease-in-out;
  transition: border-color .15s ease-in-out,box-shadow .15s ease-in-out;
  width: 100%;
  min-width: 0px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding-right: 5px;
  padding-left: 5px;
  padding-top: 4px;
  padding-bottom: 4px;
  font-size: input;
  outline: none;
  color: input;
  background-color: inputBackground;
}

.c10.c10:focus {
  --x-shadow: inputFocus;
  box-shadow: var(--x-ring-shadow,0 0 #0000),var(--x-shadow);
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <h1
          class="c3"
        >
          Next Ticker
        </h1>
        <div
          class="c4"
        >
          some text about this app
        </div>
      </div>
      <div
        class="c5"
      >
        <div
          class=""
        >
          <div
            class="c6"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="downshift-0-menu"
              class="c7"
              role="combobox"
              width="100%"
            >
              <div
                class="c8"
              >
                <div
                  class="c9"
                >
                  Symbol
                </div>
                <input
                  aria-autocomplete="list"
                  aria-controls="downshift-0-menu"
                  aria-labelledby="downshift-0-label"
                  autocomplete="off"
                  class="c10"
                  id="downshift-0-input"
                  value="AAPL"
                />
              </div>
            </div>
            <div
              class="c8"
            >
              <div
                class="c9"
              >
                Initial investment
              </div>
              <input
                class="c10"
                value="80000"
              />
            </div>
          </div>
          <div
            class="c11"
          >
            <div
              class="c8"
            >
              <div
                class="c9"
              >
                From date
              </div>
              <input
                class="c10"
                value="2020-01-01"
              />
            </div>
            <div
              class="c8"
            >
              <div
                class="c9"
              >
                To date
              </div>
              <input
                class="c10"
                value="2021-01-01"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
